apollo-router-core/src/plugins/csrf.rs

```diff
@@ -0,0 +1,289 @@
+use tower::{BoxError, ServiceBuilder, ServiceExt};
+
+#[derive(Deserialize, Debug, Clone, JsonSchema)]
+#[serde(deny_unknown_fields)]
+struct CSRFConfig {
+    #[serde(default)]
+    disabled: bool,
+}
+
+static NON_PREFLIGHTED_HEADER_NAMES: &[HeaderName] = &[
+    header::ACCEPT,
+    header::ACCEPT_LANGUAGE,
+    header::CONTENT_LANGUAGE,
+    header::CONTENT_TYPE,
+    header::RANGE,
+];
+
+static NON_PREFLIGHTED_CONTENT_TYPES: &[&str] = &[
+    "application/x-www-form-urlencoded",
+    "multipart/form-data",
+    "text/plain",
+];
+
+#[derive(Debug, Clone)]
+struct Csrf {
+    config: CSRFConfig,
+}
+
+#[async_trait::async_trait]
+impl Plugin for Csrf {
+    type Config = CSRFConfig;
+
+    async fn new(config: Self::Config) -> Result<Self, BoxError> {
+        Ok(Csrf { config })
+    }
+
+    fn router_service(
+        &mut self,
+        service: BoxService<RouterRequest, RouterResponse, BoxError>,
+    ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
+        if !self.config.disabled {
+            ServiceBuilder::new()
+                .checkpoint(move |req: RouterRequest| {
+                    if should_accept(&req) {
+                        Ok(ControlFlow::Continue(req))
+                    } else {
+                        let error = crate::Error {
+                            message: format!("This operation has been blocked as a potential Cross-Site Request Forgery (CSRF). \
+                            Please either specify a 'content-type' header (with a mime-type that is not one of {}) \
+                            or provide a header such that the request is preflighted: \
+                            https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#simple_requests", 
+                            NON_PREFLIGHTED_CONTENT_TYPES.join(",")),
+                            locations: Default::default(),
+                            path: Default::default(),
+                            extensions: Default::default(),
+                        };
+                        let res = RouterResponse::builder()
+                            .error(error)
+                            .status_code(StatusCode::BAD_REQUEST)
+                            .context(req.context)
+                            .build()?;
+                        Ok(ControlFlow::Break(res))
+                    }
+                })
+                .service(service)
+                .boxed()
+        } else {
+            service
+        }
+    }
+}
+
+fn should_accept(req: &RouterRequest) -> bool {
+    let headers = req.originating_request.headers();
+    headers_require_preflight(headers) || content_type_requires_preflight(headers)
+}
+
+fn headers_require_preflight(headers: &HeaderMap) -> bool {
+    headers
+        .keys()
+        .any(|header_name| !NON_PREFLIGHTED_HEADER_NAMES.contains(header_name))
+}
+
+fn content_type_requires_preflight(headers: &HeaderMap) -> bool {
+    headers
+        .get(header::CONTENT_TYPE)
```

Comment:
https://fetch.spec.whatwg.org/#cors-safelisted-request-header claims 
`This intentionally does not use [extract a MIME type](https://fetch.spec.whatwg.org/#concept-header-extract-mime-type) as that algorithm is rather forgiving and servers are not expected to implement it.`

The example seems to concur, so I'm tempted to:

- use `get_all`
- bail if there are more than one value
- not parse the mime type and check against string representation (as it looks like what the spec does)
wdyt?

Reply:
should be fixed in [331a5e1](https://github.com/apollographql/router/pull/1006/commits/331a5e18e6cc20858b0d0ffd28719e930d178d6b) but i ll add some tests to make sure it s the case

Reply:
@o0Ignition0o I'll look at the code in a sec, but I don't think "parse and get the essence of each value" is what I'm asking for. The fetch spec wants you to parse just one content-type — it's just that that content-type should be formed by combining all content-type headers rather than by just arbitrarily picking one of them (which is what `get()` does)

Reply:
can you please point me to the spec? I can't seem to find the "combine all values" part (and all i can find is how to parse one mime type https://mimesniff.spec.whatwg.org/#parse-a-mime-type)